### ISSUE TITLE
An existing connection was forcibly closed by the remote host; IOCP: (Busy=2,Free=998,Min=8,Max=1000), WORKER: (Busy=1,Free=32766,Min=8,Max=32767)

### ISSUE THREAD

**Issue by raviedara** [StackExchange/StackExchange.Redis#889]
Using 1.2.6 version.
Trying to use Redis cache in Acceptance test cases of 145.
It is breaking either @ FLUSHALL / GET/ DEL.
I considered not to open and close many times but still it is failing @ DEL as i need to change the key for every scenario in my acceptance testing.

Please let me know if you have any solution?

Issue Iam getting :

StackExchange.Redis.RedisConnectionException : No connection is available to service this operation: FLUSHALL; SocketClosed on srmtestservices.redis.cache.windows.net:6380/Subscription, origin: ProcessReadBytes, input-buffer: 0, outstanding: 3, last-read: 0s ago, last-write: 0s ago, unanswered-write: 161213s ago, keep-alive: 60s, pending: 0, state: ConnectedEstablishing, last-heartbeat: never, last-mbeat: -1s ago, global: 0s ago, mgr: Inactive, err: never; IOCP: (Busy=2,Free=998,Min=8,Max=1000), WORKER: (Busy=1,Free=32766,Min=8,Max=32767), Local-CPU: n/a
  ----> StackExchange.Redis.RedisConnectionException : SocketClosed on srmtestservices.redis.cache.windows.net:6380/Subscription, origin: ProcessReadBytes, input-buffer: 0, outstanding: 3, last-read: 0s ago, last-write: 0s ago, unanswered-write: 161213s ago, keep-alive: 60s, pending: 0, state: ConnectedEstablishing, last-heartbeat: never, last-mbeat: -1s ago, global: 0s ago, mgr: Inactive, err: never
     at .Acceptance.Security.ProjectRegionalManagersGroupFeature.ICanSeeTheAccessibleScreens() in \Security\Scenario09.feature:line 26
--RedisConnectionException

----------------------------

code to flush:

 private static void FlushRedisCache(bool counterNotSet = false)
        {
            if (!counterNotSet)
            {
                counter = 1;
            }

            var redisCache = ConfigurationManager.AppSettings["CacheConnectionString"];
            var options = ConfigurationOptions.Parse(redisCache);
            options.AllowAdmin = true;
            options.ConnectRetry = 3;
            var conn = ConnectionMultiplexer.Connect(options);
            var endpoints = conn.GetEndPoints(true);
            foreach (var endpoint in endpoints)
            {
                var server = conn.GetServer(endpoint);
                try
                {
                    server.FlushAllDatabases();
                }
                catch (Exception ex)
                {
                    // try again
                    if (counter <= 3)
                    {
                        counter++;
                        FlushRedisCache(true);
                    }
                    else throw ex;
                }
            }
        }

![capture](https://user-images.githubusercontent.com/18550176/42730289-95af512c-87e9-11e8-9c20-a14e2bb64e07.JPG)

**Comment by mgravell** [comment]
In the screenshot, I can see `-ERR max number of clients reached`, so: the immediate problem is that the server is rejecting you. A connection (which can be approximated to a `ConnectionMultiplexer`) should be treated as an expensive resource: you should spin it up and reuse it, sharing it between multiple code paths. If you're connecting multiple times, that's probably the cause - you would typically connect once and reuse that instance aggressively (it is thread-safe, etc, and designed for heavy concurrent load).

Also note that in the 2.0 branch we have completely reworked the socket code, but given the error message in the background, I don't this this is directly related - just : perhaps useful to know.

**Comment by raviedara** [comment]
Yes , I spin it up once and used in all test cases to remove / add the keys and after all test cases ran i again tried to close it. If that was the case after few tests ran (say 40) it was throwing issue 
StackExchange.Redis.RedisConnectionException : No connection is available to service this operation: DEL.

so do you want me to update to 2.0 version i can see that is still in alpha.

FYI
[BeforeTestRun]
        public static void BeforeTestRun()
        {
            counter = 1;
            FlushRedisCache();
        }

        [AfterTestRun]
        public static void AfterTestRun()
        {
            counter = 1;
            FlushRedisCache();
        }

and every scenario i made key DEL / ADD like below:

 public static void SetAccesibilityOfCache(UserGroup userGroup, bool counterNotSet = false)
        {
            if (!counterNotSet)
            {
                counter = 1;
            }

            try
            {
                var accessibilityName = $"AccessibilityOf_";
                FlushItem(accessibilityName);

                var task = _cache.SetAsync(accessibilityName, userGroup);
                task.Wait();
            }
            catch (RedisConnectionException ex)
            {
                if (counter <= 5)
                {
                    // try to open it back.
                    Thread.Sleep(1000);
                    FlushRedisCache();
                    _cache = new RedisCache(WebRequest.CacheUrl);
                    counter++;
                    SetAccesibilityOfCache(userGroup, true);
                }
            }
        }
![capture](https://user-images.githubusercontent.com/18550176/42733141-3f00476a-8824-11e8-8385-1c828ac66bb3.JPG)

**Comment by mgravell** [comment]
But... FlushRedisCache in your code is very clearly creating its own connection that is not disposed... If you have a reused connection, should that (FlushRedisCache) be using it? Or at least disposing the one it creates?

**Comment by raviedara** [comment]
First of all thkz for spotting the issue. 
 public void FlushAll()
        {
            var endpoints = _connection.GetEndPoints(true);
            foreach (var endpoint in endpoints)
            {
                var server = _connection.GetServer(endpoint);
                try
                {
                    server.FlushAllDatabases();
                }
                catch (Exception ex)
                {
                    throw ex;
                }
            }
            _connection.Dispose();
        }

 private static void FlushRedisCache(bool counterNotSet = false)
        {
            _cache?.FlushAll();
        }

 [AfterScenario]
        public static void AfterTestRun()
        {
            FlushRedisCache();
            var isFound = FeatureContext.Current?.FeatureInfo?.Tags.Contains("clearAllTables");
            if (isFound.HasValue && isFound.Value)
            {
                ClearAllTables();
            }
        }
Chnged the code as above. 

But the issue i mentioned above is still there if I put my FlushRedisCache() method in AfterTestRun method. So for now I kept in AfterScenario and all my test cases (145) are working in a single run.

I know i am consuming lot of resource by disposing cache after every scenario and i dont want to do that. Can you please suggest me how to proceed on this ?

**Comment by mgravell** [comment]
Quick fix: make sure you dispose it when you're done with whatever commands
you are issuing. For a standalone method like this, a "using" block should
work. More generally, consider whether you can simply keep the same
multiplexer instance around for the duration of your tests and reuse it for
everything including the flush.
